### PR TITLE
[compact]support compact unaware bucket for combine_mode

### LINKED ISSUE #2670: [Feature]  Whole database compression combined mode does not support append unware bucket table
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Whole database compression combined mode does not support append unware bucket table。
f7467db14a7d8d34906f776d5695c4e3

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by wg1026688210**
Refact the compaction of combine mode for supporting compacting unaware bucket table 

### Purpose

Linked issue: close #2670 

### Tests

[CompactDatabaseActionITCase.java](https://github.com/apache/incubator-paimon/pull/2858/files#diff-9eae358dbfd2474f886992026b8840e7a0d2438bef74eae239fa98974462287c) for IT case in stream and batch mode
CompactionTaskSerializerTest.java for the test of serializing compaction task
AppendOnlyMultiTableCompactionWorkerOperatorTest.java for the  test of  AppendOnlyMultiTableCompactionWorkerOperator

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CombinedUmawareCompactionWorkerOperator.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    public CombinedUmawareCompactionWorkerOperator(
+            Catalog.Loader catalogLoader, String commitUser, Options options) {
+        super(options);
+        this.commitUser = commitUser;
+        this.catalogLoader = catalogLoader;
+    }
+
+    @VisibleForTesting
+    Iterable<Future<CommitMessage>> result() {
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CombinedUmawareCompactionWorkerOperator.java:79]: Is there any place using this method?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/DividedUnawareCompactionWorkerOperator.java [hunk 2]
```diff
@@ -20,92 +20,68 @@
...
 
     private static final Logger LOG =
-            LoggerFactory.getLogger(AppendOnlyTableCompactionWorkerOperator.class);
+            LoggerFactory.getLogger(DividedUnawareCompactionWorkerOperator.class);
 
     private final FileStoreTable table;
     private final String commitUser;
 
-    private transient AppendOnlyFileStoreWrite write;
+    private UnwareBucketCompactionHelper compactionHelper;
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/DividedUnawareCompactionWorkerOperator.java:54]: transient?

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyCompactionTask.java [hunk 3]
```diff
@@ -39,13 +40,22 @@ public class AppendOnlyCompactionTask {
...
     private final List<DataFileMeta> compactAfter;
 
+    private final Identifier tableIdentifier;
+
     public AppendOnlyCompactionTask(BinaryRow partition, List<DataFileMeta> files) {
+        this(partition, files, Identifier.EMPTY);
+    }
+
+    public AppendOnlyCompactionTask(
+            BinaryRow partition, List<DataFileMeta> files, Identifier identifier) {
```
Comment (leaves12138) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyCompactionTask.java:50]: I think a new class `MultiTableAppendOnlyCompactionTask` which extends `AppendOnlyCompactionTask` seems better, in this way, we don't need Identifier.EMPTY which may cause hidden bug.

Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyCompactionTask.java:50]: We can just put string in `MultiTableAppendOnlyCompactionTask`, then we can remove `IdentifierSerializer`.

**Review by leaves12138 — COMMENTED** [review]
You can remove `MultiTablesCompactorSourceFunction` if you have renamed it

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/AbstractBucketScanLogic.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+ * buckets during compaction.
+ *
+ * @param <T> the result of scanning file :
+ *     <ol>
+ *       <li>{@link Split} for the table with multi buckets, such as dynamic or fixed bucket table.
+ *       <li>{@link AppendOnlyCompactionTask} for the table witch fixed single bucket ,such as
+ *           unaware bucket table.
+ *     </ol>
+ */
+public abstract class AbstractBucketScanLogic<T> {
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/AbstractBucketScanLogic.java:49]: Maybe another name? Like `MultiTableScanBase`

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiBucketScanLogic.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.flink.utils.MultiTablesCompactorUtil.compactOptions;
+
+/**
+ * This class is responsible for implementing the scanning logic {@link AbstractBucketScanLogic} for
+ * the table with multi bucket such as dynamic or fixed bucket table.
+ */
+public class MultiBucketScanLogic extends AbstractBucketScanLogic<Tuple2<Split, String>> {
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiBucketScanLogic.java:49]: `MultiAwareBucketTableScan`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnwareBucketScanLogic.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Pattern;
+
+/**
+ * This class is responsible for implementing the scanning logic {@link AbstractBucketScanLogic} for
+ * the table with fix single bucket such as unaware bucket table.
+ */
+public class UnwareBucketScanLogic extends AbstractBucketScanLogic<AppendOnlyCompactionTask> {
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnwareBucketScanLogic.java:44]: `MultiUnawareBuckeTableScan`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/CompactionFileScanner.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @param <T> the result of scanning file :
+ *     <ol>
+ *       <li>{@link Split} for the table with multi buckets, such as dynamic or fixed bucket table.
+ *       <li>{@link AppendOnlyCompactionTask} for the table witch fixed single bucket ,such as
+ *           unaware bucket table.
+ *     </ol>
+ */
+public abstract class CompactionFileScanner<T> {
+    protected final AtomicBoolean isRunning;
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/CompactionFileScanner.java:39]: 可以在 CompactionFileScanner 和 AtomicBoolean 之间加个换行

Comment (leaves12138) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/CompactionFileScanner.java:39]: 其他的类也一样

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/StreamingFileScanner.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            AtomicBoolean isRunning) {
+        super(isRunning, tableScanLogic);
+        this.monitorInterval = monitorInterval;
+    }
+
+    @SuppressWarnings("BusyWait")
+    @Override
+    public void scan(SourceFunction.SourceContext<T> ctx) throws Exception {
+        while (isRunning.get()) {
+            Boolean isEmpty = tableScanLogic.collectFiles(ctx);
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/StreamingFileScanner.java:45]: Kind of weird here, use 'null' to tag 'end', can we find a better way?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CombinedUmawareCompactionWorkerOperator.java (hunk 1 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CombinedUmawareCompactionWorkerOperator.java:79]: Done ,this has  been removed.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/DividedUnawareCompactionWorkerOperator.java (hunk 2 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/DividedUnawareCompactionWorkerOperator.java:54]: done

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyCompactionTask.java (hunk 3 above)
Comment (wg1026688210) [reply paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyCompactionTask.java:50]: done ,this change has been finished at #3174

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CombinedUmawareCompactionWorkerOperator.java (hunk 1 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/CombinedUmawareCompactionWorkerOperator.java:79]: this has been finished at https://github.com/apache/paimon/pull/3175

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/AbstractBucketScanLogic.java (hunk 4 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/AbstractBucketScanLogic.java:49]: done ,this has been finished at #3179

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiBucketScanLogic.java (hunk 5 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiBucketScanLogic.java:49]: done ,this has been finished at https://github.com/apache/paimon/pull/3179

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnwareBucketScanLogic.java (hunk 6 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnwareBucketScanLogic.java:44]: done ,this has been finished at https://github.com/apache/paimon/pull/3179

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/CompactionFileScanner.java (hunk 7 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/CompactionFileScanner.java:39]: done ,this has been finished at https://github.com/apache/paimon/pull/3179

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/StreamingFileScanner.java (hunk 8 above)
Comment (wg1026688210) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/StreamingFileScanner.java:45]: I add a Simple enum for a more clear expression at #3179

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyMultiTableCompactionWorkerOperator.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        if (lazyCompactExecutor != null) {
+            // ignore runnable tasks in queue
+            lazyCompactExecutor.shutdownNow();
+            if (!lazyCompactExecutor.awaitTermination(120, TimeUnit.SECONDS)) {
+                LOG.warn(
+                        "Executors shutdown timeout, there may be some files aren't deleted correctly");
+            }
+
+            for (UnawareBucketCompactor compactor : compactorContainer.values()) {
+                compactor.close();
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/AppendOnlyMultiTableCompactionWorkerOperator.java:140]: We should consider exception.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedAwareBatchSourceFunction.java [hunk 10]
```diff
@@ -39,65 +39,53 @@
...
-                                scan.plan().splits().stream()
-                                        .map(split -> new Tuple2<>(split, identifier.getFullName()))
-                                        .collect(Collectors.toList()));
-                    }
-
-                    isEmpty = splits.isEmpty();
-                    splits.forEach(ctx::collect);
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        tableScanLogic =
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedAwareBatchSourceFunction.java:66]: just tableScan?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedAwareBatchSourceFunction.java [hunk 11]
```diff
@@ -39,65 +39,53 @@
...
-public class MultiTablesBatchCompactorSourceFunction extends MultiTablesCompactorSourceFunction {
+import static org.apache.paimon.flink.compact.MultiTableScanBase.ScanResult.FINISHED;
+import static org.apache.paimon.flink.compact.MultiTableScanBase.ScanResult.IS_EMPTY;
 
-    private static final Logger LOG =
-            LoggerFactory.getLogger(MultiTablesBatchCompactorSourceFunction.class);
+/** It is responsible for monitoring compactor source of aware bucket table in batch mode. */
+public class CombinedAwareBatchSourceFunction
+        extends CombinedCompactorSourceFunction<Tuple2<Split, String>> {
+    private static final Logger LOGGER =
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedAwareBatchSourceFunction.java:50]: line feed

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedAwareStreamingSourceFunction.java [hunk 12]
```diff
@@ -19,88 +19,68 @@
...
-                                scan.plan().splits().stream()
-                                        .map(split -> new Tuple2<>(split, identifier.getFullName()))
-                                        .collect(Collectors.toList()));
-                    }
-
-                    isEmpty = splits.isEmpty();
-                    splits.forEach(ctx::collect);
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        tableScanLogic =
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedAwareStreamingSourceFunction.java:65]: just tableScan

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareBatchSourceFunction.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+import static org.apache.paimon.flink.compact.MultiTableScanBase.ScanResult.FINISHED;
+import static org.apache.paimon.flink.compact.MultiTableScanBase.ScanResult.IS_EMPTY;
+
+/**
+ * It is responsible for the batch compactor source of the table with unaware bucket in combined
+ * mode.
+ */
+public class CombinedUnawareBatchSourceFunction
+        extends CombinedCompactorSourceFunction<MultiTableAppendOnlyCompactionTask> {
+    private static final Logger LOGGER =
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareBatchSourceFunction.java:50]: line feed

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareBatchSourceFunction.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            Pattern includingPattern,
+            Pattern excludingPattern,
+            Pattern databasePattern) {
+        super(catalogLoader, includingPattern, excludingPattern, databasePattern, false);
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        tableScanLogic =
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareBatchSourceFunction.java:65]: ditto

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareStreamingSourceFunction.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+import static org.apache.paimon.flink.compact.MultiTableScanBase.ScanResult.FINISHED;
+import static org.apache.paimon.flink.compact.MultiTableScanBase.ScanResult.IS_EMPTY;
+
+/**
+ * It is responsible for monitoring compactor source in stream mode for the table of unaware bucket.
+ */
+public class CombinedUnawareStreamingSourceFunction
+        extends CombinedCompactorSourceFunction<MultiTableAppendOnlyCompactionTask> {
+    private final long monitorInterval;
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareStreamingSourceFunction.java:44]: line feed

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareStreamingSourceFunction.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+            Pattern databasePattern,
+            long monitorInterval) {
+        super(catalogLoader, includingPattern, excludingPattern, databasePattern, true);
+        this.monitorInterval = monitorInterval;
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+        tableScanLogic =
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/CombinedUnawareStreamingSourceFunction.java:60]: ditto

File: paimon-core/src/main/java/org/apache/paimon/append/MultiTableAppendOnlyCompactionTask.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.paimon.catalog.Identifier;
+import org.apache.paimon.data.BinaryRow;
+import org.apache.paimon.io.DataFileMeta;
+
+import java.util.List;
+import java.util.Objects;
+
+/** Compaction task for multi table . */
+public class MultiTableAppendOnlyCompactionTask extends AppendOnlyCompactionTask {
+    private final Identifier tableIdentifier;
```
Comment (leaves12138) [inline paimon-core/src/main/java/org/apache/paimon/append/MultiTableAppendOnlyCompactionTask.java:30]: line feed

File: paimon-core/src/main/java/org/apache/paimon/table/sink/MultiTableCompactionTaskSerializer.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.paimon.utils.SerializationUtils.deserializeBinaryRow;
+import static org.apache.paimon.utils.SerializationUtils.serializeBinaryRow;
+
+/** Serializer for {@link MultiTableAppendOnlyCompactionTask}. */
+public class MultiTableCompactionTaskSerializer
+        implements VersionedSerializer<MultiTableAppendOnlyCompactionTask> {
+    private static final int CURRENT_VERSION = 1;
```
Comment (leaves12138) [inline paimon-core/src/main/java/org/apache/paimon/table/sink/MultiTableCompactionTaskSerializer.java:41]: line feed

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiAwareBucketTableScan.java [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.flink.utils.MultiTablesCompactorUtil.compactOptions;
+
+/**
+ * This class is responsible for implementing the scanning logic {@link MultiTableScanBase} for the
+ * table with multi bucket such as dynamic or fixed bucket table.
+ */
+public class MultiAwareBucketTableScan extends MultiTableScanBase<Tuple2<Split, String>> {
+    protected transient Map<Identifier, BucketsTable> tablesMap;
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiAwareBucketTableScan.java:46]: line feed (new line)

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiTableScanBase.java [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+ * @param <T> the result of scanning file :
+ *     <ol>
+ *       <li>Tuple2<{@link Split},String> for the table with multi buckets, such as dynamic or fixed
+ *           bucket table.
+ *       <li>{@link MultiTableAppendOnlyCompactionTask} for the table witch fixed single bucket
+ *           ,such as unaware bucket table.
+ *     </ol>
+ */
+public abstract class MultiTableScanBase<T> {
+    private static final Logger LOG = LoggerFactory.getLogger(MultiTableScanBase.class);
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/MultiTableScanBase.java:52]: line feed

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactor.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/** The Compactor of unaware bucket table to execute {@link AppendOnlyCompactionTask}. */
+public class UnawareBucketCompactor {
+    private final FileStoreTable table;
```
Comment (leaves12138) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactor.java:41]: line feed

**Review by leaves12138 — APPROVED** [review]
+1

**Pull request merged**